src/Microsoft.Data.Analysis/DataFrame.IO.cs

```diff
@@ -269,54 +272,54 @@ private static DataFrameColumn CreateColumn(Type kind, string[] columnNames, int
                 }
                 columns.Insert(0, indexColumn);
             }
+
+            textReader.Dispose();
             return ret;
         }
 
-        private class CsvLines : IEnumerable<string>
+        private class WrappedStreamReaderOrStringReader
         {
-            private CsvLineEnumerator enumerator;
-            public CsvLines(CsvLineEnumerator csvLineEnumerator)
+            private Stream _stream;
+            private long _initialPosition;
+            private Encoding _encoding;
+            private string _csvString;
+
+            public WrappedStreamReaderOrStringReader(Stream stream, Encoding encoding)
             {
-                enumerator = csvLineEnumerator;
+                _stream = stream;
+                _initialPosition = stream.Position;
+                _encoding = encoding;
+                _csvString = null;
             }
 
-            public IEnumerator<string> GetEnumerator() => enumerator;
-
-            IEnumerator IEnumerable.GetEnumerator() => enumerator;
-        }
-
-        private class CsvLineEnumerator : IEnumerator<string>
-        {
-            private StreamReader streamReader;
-            private string currentLine;
-            private long streamStartPosition;
-            public CsvLineEnumerator(StreamReader csvStream)
+            public WrappedStreamReaderOrStringReader(string csvString)
             {
-                streamStartPosition = csvStream.BaseStream.Position;
-                streamReader = csvStream;
-                currentLine = null;
+                _csvString = csvString;
+                _initialPosition = 0;
+                _encoding = null;
+                _stream = null;
             }
 
-            public string Current => currentLine;
-
-            object IEnumerator.Current => currentLine;
+            public long InitialPosition => _initialPosition;
 
-            public void Dispose()
+            // Returns a new TextReader. If the wrapped object is a stream, the stream is reset to its initial position. 
+            public TextReader TextReader
```

Comment:
This is a bit odd usage of a property. I think this should be a method instead, since typically properties should behave more like fields.